text/0034-precomputed-entity-attributes.md

```diff
@@ -0,0 +1,186 @@
+  it splits out the changes from RFC 28 that relate to precomputed entity
+  attributes, without the larger changes in RFC 28 regarding `EntityDataSource`.
+- This RFC is also the successor to [RFC 23](https://github.com/cedar-policy/rfcs/pull/23).
+
+## Timeline
+
+- Started: 2023-10-24 (predecessor [#28](https://github.com/cedar-policy/rfcs/pull/28) started on 2023-10-03, and predecessor [#23](https://github.com/cedar-policy/rfcs/pull/23) started on 2023-07-18)
+- Entered FCP:
+- Accepted:
+- Landed:
+
+## Summary
+
+Internally, Cedar entity objects will store their attribute values as
+precomputed `Value`s, rather than in `RestrictedExpression` form.
+This entails some minor breaking changes to the Cedar public API, but provides
+important efficiency gains.
+
+## Motivation
+
+1. Today, every `is_authorized()` request re-evaluates all of the attributes in the
+`Entities`; they are stored as restricted expressions and evaluated to `Value`
+on every call to `is_authorized()`. (See
+[cedar#227](https://github.com/cedar-policy/cedar/issues/227).)
+This is inefficient.
+
+2. Today, `Entities::evaluate()` is implemented on `main` and provides a way to
+mitigate this issue, but it is opt-in and we'd rather have it be the default
+behavior.
+Notably, as of this writing, `Entities::evaluate()` is not released yet, so we
+can make changes in this area without breaking our users.
+
+3. Today, `Entity::new()` requires callers to pass in attribute values as
+`RestrictedExpression`.
+For some callers, evaluating these `RestrictedExpression`s is an unnecessary
+source of runtime evaluation errors (and performance overhead).
+
+## Detailed design
+
+This RFC has three components:
+
+### Component 1: Store attributes as precomputed `Value`s
+
+First, we redefine `Entities` and `Entity` to hold attributes as `Value`s
+instead of as restricted expressions.
+This addresses the first point in [Motivation](#motivation).
+
+This change is breaking for the public API in two (small) ways:
+- `Entity::new()`, `Entities::from_json_*()`, and friends can take the same
+inputs, but will need to sometimes return an evaluation error due to errors
+evaluating the entity attributes (particularly, errors from evaluating extension
+functions for attribute values of type `ipaddr` or `decimal`).
+Today, `Entity::new()` never returns an error, and the `Entities` functions can
+return `EntitiesError` but not an evaluation error.
+- Conversely, `Entity::attr()` need not return an error anymore, as all of the
+attributes will already be evaluated. (If we are very concerned about backward
+compatibility, we could keep the current signature and just never return `Err`.
+Currently, this RFC is proposing we change the signature to not contain
+`Result`, since we're already making the related breaking change to
+`Entity::new()`.)
+
+Accepting these breaking changes allows us to give users the best-performing
+behavior by default.
```

Comment:
What will happen to customer code when they upgraded? I think they will get a type error, which they can easily fix, is that right?

Reply:
Yes, users will get type errors (or other compile errors) for all the breaking changes contemplated in this RFC.  As far as "easily fix", existing users of `EvalResult` (if there are any) will have the most involved fixes, as the underlying structure of `EvalResult` will have changed (if we take the third option in Component 2, as currently proposed).  The fixes for the changes in Component 1 should be easy for users. (and Component 3 is not a breaking change for any users on a released Cedar version -- and even if someone is on a pre-release version of Cedar and is using `.evaluate()`, they simply have to remove it.)

Reply:
Now that the RFC has been updated to remove Component 2 (leaving it in Alternative A), the only breaking changes that remain are the ones in Component 1, which should be easy for users (and yes, surface as type errors or other compile errors).